compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  gen += this->genFuncNumericID(fn);
+  gen += "(void)";
+  gen += scope_begin;
+
+  gen += this->genServersideRPC(fn);
+  gen += "return 0;\n";
+ 
+  gen += scope_end;
+  gen += "\n";
+
+  this->setOutputAndWrite(&this->fiServerBundle, gen);
+ 
+  return;
+}
+
+std::string MLIContext::genDebugPrintCall(FnSymbol* fn, const char* pfx) {
+  std::string msg;
+
+  msg += "Calling: ";
+  msg += toString(fn);
+
+  return this->genDebugPrintCall(msg.c_str(), pfx);
+}
+
+std::string MLIContext::genDebugPrintCall(const char* msg, const char* pfx) {
+  std::string gen;
+
+  gen += "printf(\"";
+
+  if (pfx && pfx != "") {
+    gen += pfx;
+    gen += " ";
+  }
+
+  gen += "%%s\\n\", \"";
+  gen += msg;
+  gen += "\");\n";
+
+  return gen;
+}
+
+std::string MLIContext::genFuncNumericID(FnSymbol* fn) {
+  return std::to_string((int64_t) fn->id);
+}
+
+std::string MLIContext::genServerWrapperCall(FnSymbol* fn) {
+  std::string gen;
+
+  gen += "chpl_mli_swrapper_";
+  gen += this->genFuncNumericID(fn);
+  gen += "();\n";
+
+  return gen;
+}
+  
+std::string
+MLIContext::genServerDispatchSwitch(const std::vector<FnSymbol*>& fns) {
+  std::string gen;
+
+  gen += "int64_t chpl_mli_sdispatch";
+  gen += "(int64_t function)";
+  gen += scope_begin;
+  gen += this->genNewDecl("int", "err");
+  gen += "switch (function)";
+  gen += scope_begin;
+
+  for_vector(FnSymbol, fn, fns) {
+    if (not fn->hasFlag(FLAG_EXPORT) || fn->hasFlag(FLAG_GEN_MAIN_FUNC)) {
+      continue;
+    }
+
+    gen += "case ";
+    gen += this->genFuncNumericID(fn);
+    gen += ": ";
+
+    gen += scope_begin;
+    
+    if (this->injectDebugPrintlines) {
+      gen += this->genDebugPrintCall(fn, "[Server]");
+    }
+    
+    gen += "err = ";
+    gen += this->genServerWrapperCall(fn);
+    gen += scope_end;
+    gen += "break;\n";
+  }
+
+  gen += "default: return CHPL_MLI_ERROR_NOFUNC; break;\n";
+  gen += scope_end;
+  gen += "return err;\n";
+  gen += scope_end;
+
+  return gen;
+}
+
+//
+// TODO: This is unused right now (no struct type support yet).
+//
+bool MLIContext::structContainsOnlyPrimitiveScalars(Type *t) {
+  return false;
+}
+
+//
+// TODO: This filter will change as we support more and more type classes.
+//  - [ ] dtStringC
+//  - [ ] Chapel `string` type
+//  - [ ] ???
+//
+bool MLIContext::isSupportedType(Type* t) {
+  return (isPrimitiveScalar(t));
+}
+
+void MLIContext::verifyPrototype(FnSymbol* fn) {
+
+  if (fn->retType != dtVoid && not isSupportedType(fn->retType)) {
+    USR_FATAL("MLI does not support code generation for type", fn->retType);
+  }
+
+  // Loop through all formals and error gen if a type is unsupported.
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+    if (isSupportedType(as->type)) { continue; }
+    USR_FATAL("MLI does not support code generation for type", as->type);
+  }
+
+  return;
+}
+
+Type* MLIContext::getTypeFromFormal(ArgSymbol* as) {
+  if (as == NULL) { return NULL; }
```

Comment:
Uhh, it seems like things would go very wrong in this case.  I think you should use INT_FATAL instead of just returning NULL back

Reply:
Yeah, I figured that the ArgSymbol code already had internal asserts to avoid that, but It would probably be better to use one myself.

Reply:
But this doesn't assert anything.  If the arg is NULL, you'll just output a partial line [here](https://github.com/chapel-lang/chapel/pull/12893/files/44a0d68999a84709ed57f3c08b62b15a03c4cc1f#diff-10fd9d3c5308335749c462da28631d6bR714) and that will break compilation in a weird way when we call the C compiler, right?

Reply:
I added an internal assert instead.